olp-cpp-sdk-dataservice-read/tests/StreamLayerClientImplTest.cpp

```diff
@@ -0,0 +1,303 @@
+
+constexpr auto kUrlStreamSubscribe =
+    R"(https://some.stream.url/stream/v2/catalogs/hrn:here:data::olp-here-test:hereos-internal-test-v2/layers/testlayer/subscribe?mode=serial)";
+
+constexpr auto kHttpResponseLookupStream =
+    R"jsonString([{"api":"stream","version":"v2","baseURL":"https://some.stream.url/stream/v2/catalogs/hrn:here:data::olp-here-test:hereos-internal-test-v2","parameters":{}}])jsonString";
+
+constexpr auto kHttpResponseSubscribe =
+    R"jsonString({ "nodeBaseURL": "https://some.stream.url/stream/v2/catalogs/hrn:here:data::olp-here-test:hereos-internal-test-v2", "subscriptionId": "12345" })jsonString";
+
+constexpr auto kHttpResponseSubscribeForbidden =
+    R"jsonString({ "error": "Forbidden Error", "error_description": "Error description" })jsonString";
+
+const std::string kCatalog =
+    "hrn:here:data::olp-here-test:hereos-internal-test-v2";
+const auto kHRN = olp::client::HRN::FromString(kCatalog);
+const std::string kLayerId = "testlayer";
+const auto kTimeout = std::chrono::seconds(5);
+const auto kSubcriptionId = "12345";
+
+enum class RequestMethod { GET, POST };
+
+template <class T>
+void SetupNetworkExpectation(NetworkMock& network_mock, T url, T response,
+                             int status,
+                             RequestMethod method = RequestMethod::GET) {
+  switch (method) {
+    case RequestMethod::GET: {
+      EXPECT_CALL(network_mock, Send(IsGetRequest(url), _, _, _, _))
+          .WillOnce(ReturnHttpResponse(
+              olp::http::NetworkResponse().WithStatus(status), response));
+      return;
+    }
+    case RequestMethod::POST: {
+      EXPECT_CALL(network_mock, Send(IsPostRequest(url), _, _, _, _))
+          .WillOnce(ReturnHttpResponse(
+              olp::http::NetworkResponse().WithStatus(status), response));
+      return;
+    }
+    default:
+      return;
+  }
+}
+
+TEST(StreamLayerClientImplTest, Subsribe) {
+  auto network_mock = std::make_shared<NetworkMock>();
+  auto cache_mock = std::make_shared<CacheMock>();
+  olp::client::OlpClientSettings settings;
+  settings.network_request_handler = network_mock;
+  settings.cache = cache_mock;
+
+  {
+    SCOPED_TRACE("Subscribe success");
+
+    StreamLayerClientImpl client(kHRN, kLayerId, settings);
+
+    SetupNetworkExpectation(*network_mock, kUrlLookupStream,
+                            kHttpResponseLookupStream,
+                            olp::http::HttpStatusCode::OK);
+
+    SetupNetworkExpectation(
+        *network_mock, kUrlStreamSubscribe, kHttpResponseSubscribe,
+        olp::http::HttpStatusCode::CREATED, RequestMethod::POST);
+
+    std::promise<SubscribeResponse> promise;
+    std::future<SubscribeResponse> future(promise.get_future());
+
+    auto token = client.Subscribe(
+        SubscribeRequest(),
+        [&](SubscribeResponse response) { promise.set_value(response); });
+
+    EXPECT_EQ(future.wait_for(kTimeout), std::future_status::ready);
```

Comment:
```suggestion
    ASSERT_EQ(future.wait_for(kTimeout), std::future_status::ready);
```
I think you should use assert here in order to protect yourself from infinite waiting on future's result